src/main/java/com/puppycrawl/tools/checkstyle/site/XdocsTemplateParser.java

```diff
@@ -0,0 +1,213 @@
+///////////////////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.site;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.Reader;
+import java.io.StringReader;
+import java.io.StringWriter;
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+
+import javax.swing.text.html.HTML.Attribute;
+
+import org.apache.maven.doxia.macro.MacroExecutionException;
+import org.apache.maven.doxia.macro.MacroRequest;
+import org.apache.maven.doxia.macro.manager.MacroNotFoundException;
+import org.apache.maven.doxia.module.xdoc.XdocParser;
+import org.apache.maven.doxia.parser.ParseException;
+import org.apache.maven.doxia.parser.Parser;
+import org.apache.maven.doxia.sink.Sink;
+import org.codehaus.plexus.component.annotations.Component;
+import org.codehaus.plexus.util.IOUtil;
+import org.codehaus.plexus.util.xml.pull.XmlPullParser;
+
+/**
+ * Parser for Checkstyle's xdoc templates.
+ * This parser is responsible for generating xdocs({@code .xml}) from the xdoc
+ * templates({@code .xml.template}). The templates are regular xdocs with custom
+ * macros for generating dynamic content - properties, examples, etc.
+ * This parser behaves just like the {@link XdocParser} with the difference that all
+ * elements apart from the {@code macro} element are copied as is to the output.
+ *
+ * @see ExampleMacro
+ */
+@Component(role = Parser.class, hint = "xdocs-template")
+public class XdocsTemplateParser extends XdocParser {
+
+    /** The source content of the input reader. Used to pass into macros. */
+    private String sourceContent;
+
+    /** A macro name. */
+    private String macroName;
+
+    /** The macro parameters. */
+    private Map<String, Object> macroParameters = new HashMap<>();
+
+    @Override
+    public void parse(Reader source, Sink sink, String reference) throws ParseException {
+        sourceContent = null;
+
+        try (Reader reader = source) {
+            final StringWriter contentWriter = new StringWriter();
+            IOUtil.copy(reader, contentWriter);
+            sourceContent = contentWriter.toString();
+            super.parse(new StringReader(sourceContent), sink, reference);
+        }
+        catch (IOException ioException) {
+            throw new ParseException("Error reading the input source", ioException);
+        }
+        finally {
+            sourceContent = null;
+        }
+    }
+
+    @Override
+    protected void handleStartTag(XmlPullParser parser, Sink sink) throws MacroExecutionException {
+        if (parser.getName().equals(DOCUMENT_TAG.toString())) {
+            sink.body();
+            sink.rawText(parser.getText());
+        }
+        else if (parser.getName().equals(MACRO_TAG.toString())) {
+            processMacroStart(parser);
+            setIgnorableWhitespace(true);
+        }
+        else if (parser.getName().equals(PARAM.toString())) {
```

Comment:
```suggestion
        else if (parserName.equals(PARAM.toString()) && isSecondParsing()) {
```

Reply:
Also please extract `parser.getname()` to a variable, I think it is good to do this whenever we call a method more than once.

Reply:
Done for both. Thanks for the clarification.